File: pkg/dev_compiler/test/codegen_expected/language/function_subtype1_test.js
dart_library.library('language/function_subtype1_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__function_subtype1_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const function_subtype1_test = Object.create(null);
  let C = () => (C = dart.constFn(function_subtype1_test.C$()))();
  let COfint_ = () => (COfint_ = dart.constFn(function_subtype1_test.C$(function_subtype1_test.int_)))();
  let COfFunction = () => (COfFunction = dart.constFn(function_subtype1_test.C$(core.Function)))();
  let COf_ = () => (COf_ = dart.constFn(function_subtype1_test.C$(function_subtype1_test._)))();
  let COfvoid_ = () => (COfvoid_ = dart.constFn(function_subtype1_test.C$(function_subtype1_test.void_)))();
  let COfObject_ = () => (COfObject_ = dart.constFn(function_subtype1_test.C$(function_subtype1_test.Object_)))();
  let COfdouble_ = () => (COfdouble_ = dart.constFn(function_subtype1_test.C$(function_subtype1_test.double_)))();
  let COfvoid__int = () => (COfvoid__int = dart.constFn(function_subtype1_test.C$(function_subtype1_test.void__int)))();
  let COfint__int = () => (COfint__int = dart.constFn(function_subtype1_test.C$(function_subtype1_test.int__int)))();
  let COfint__Object = () => (COfint__Object = dart.constFn(function_subtype1_test.C$(function_subtype1_test.int__Object)))();
  let COfObject__int = () => (COfObject__int = dart.constFn(function_subtype1_test.C$(function_subtype1_test.Object__int)))();
  let COfint__double = () => (COfint__double = dart.constFn(function_subtype1_test.C$(function_subtype1_test.int__double)))();
  let COfint__int_int = () => (COfint__int_int = dart.constFn(function_subtype1_test.C$(function_subtype1_test.int__int_int)))();
  let COfinline_void_ = () => (COfinline_void_ = dart.constFn(function_subtype1_test.C$(function_subtype1_test.inline_void_)))();
  let COfinline_void__int = () => (COfinline_void__int = dart.constFn(function_subtype1_test.C$(function_subtype1_test.inline_void__int)))();
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  function_subtype1_test.C$ = dart.generic(T => {
    class C extends core.Object {}
    dart.addTypeTests(C);
    return C;
  });
  function_subtype1_test.C = C();
  function_subtype1_test._ = dart.typedef('_', () => dart.functionType(dart.dynamic, []));
  function_subtype1_test.void_ = dart.typedef('void_', () => dart.functionType(dart.void, []));
  function_subtype1_test.void_2 = dart.typedef('void_2', () => dart.functionType(dart.void, []));
  function_subtype1_test.int_ = dart.typedef('int_', () => dart.functionType(core.int, []));
  function_subtype1_test.int_2 = dart.typedef('int_2', () => dart.functionType(core.int, []));
  function_subtype1_test.Object_ = dart.typedef('Object_', () => dart.functionType(core.Object, []));
  function_subtype1_test.double_ = dart.typedef('double_', () => dart.functionType(core.double, []));
  function_subtype1_test.void__int = dart.typedef('void__int', () => dart.functionType(dart.void, [core.int]));
  function_subtype1_test.int__int = dart.typedef('int__int', () => dart.functionType(core.int, [core.int]));
  function_subtype1_test.int__int2 = dart.typedef('int__int2', () => dart.functionType(core.int, [core.int]));
  function_subtype1_test.int__Object = dart.typedef('int__Object', () => dart.functionType(core.int, [core.Object]));
  function_subtype1_test.Object__int = dart.typedef('Object__int', () => dart.functionType(core.Object, [core.int]));
  function_subtype1_test.int__double = dart.typedef('int__double', () => dart.functionType(core.int, [core.double]));
  function_subtype1_test.int__int_int = dart.typedef('int__int_int', () => dart.functionType(core.int, [core.int, core.int]));
  function_subtype1_test.inline_void_ = dart.typedef('inline_void_', () => dart.functionType(dart.void, [dart.functionType(dart.void, [])]));
  function_subtype1_test.inline_void__int = dart.typedef('inline_void__int', () => dart.functionType(dart.void, [dart.functionType(dart.void, [core.int])]));
  function_subtype1_test.main = function() {
    expect$.Expect.isTrue(COfFunction().is(new (COfint_())()));
    expect$.Expect.isFalse(COfint_().is(new (COfFunction())()));
    expect$.Expect.isTrue(COf_().is(new (COf_())()));
    expect$.Expect.isTrue(COfvoid_().is(new (COf_())()));
    expect$.Expect.isTrue(COf_().is(new (COfvoid_())()));
    expect$.Expect.isTrue(COfvoid_().is(new (COfint_())()));
    expect$.Expect.isFalse(COfint_().is(new (COfvoid_())()));
    expect$.Expect.isTrue(COfvoid_().is(new (COfvoid_())()));
    expect$.Expect.isTrue(COfint_().is(new (COfint_())()));
    expect$.Expect.isTrue(COfObject_().is(new (COfint_())()));
    expect$.Expect.isFalse(COfdouble_().is(new (COfint_())()));
    expect$.Expect.isFalse(COfvoid__int().is(new (COfint_())()));
    expect$.Expect.isFalse(COfint__int().is(new (COfvoid_())()));
    expect$.Expect.isFalse(COfvoid__int().is(new (COfvoid_())()));
    expect$.Expect.isTrue(COfint__int().is(new (COfint__int())()));
    expect$.Expect.isTrue(COfObject__int().is(new (COfint__Object())()));
    expect$.Expect.isFalse(COfint__double().is(new (COfint__int())()));
    expect$.Expect.isFalse(COfint__int().is(new (COfint_())()));
    expect$.Expect.isFalse(COfint__int_int().is(new (COfint__int())()));
    expect$.Expect.isFalse(COfint__int().is(new (COfint__int_int())()));
    expect$.Expect.isFalse(COfinline_void__int().is(new (COfinline_void_())()));
    expect$.Expect.isFalse(COfinline_void_().is(new (COfinline_void__int())()));
  };
  dart.fn(function_subtype1_test.main, VoidTodynamic());
  // Exports:
  exports.function_subtype1_test = function_subtype1_test;
});

File: tests/language_2/function_subtype/function_subtype1_test.dart
// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.
// Dart test program for constructors and initializers.

// Check function subtyping.

import 'package:expect/expect.dart';

class C<T> {}

typedef _();
typedef void void_();
typedef void void_2();
typedef int int_();
typedef int int_2();
typedef Object Object_();
typedef double double_();
typedef void void__int(int i);
typedef int int__int(int i);
typedef int int__int2(int i);
typedef int int__Object(Object o);
typedef Object Object__int(int i);
typedef int int__double(double d);
typedef int int__int_int(int i1, int i2);
typedef void inline_void_(void f());
typedef void inline_void__int(void f(int i));

main() {
  // () -> int <: Function
  Expect.isTrue(new C<int_>() is C<Function>);
  // Function <: () -> int
  Expect.isFalse(new C<Function>() is C<int_>);
  // () -> dynamic <: () -> dynamic
  Expect.isTrue(new C<_>() is C<_>);
  // () -> dynamic <: () -> void
  Expect.isTrue(new C<_>() is C<void_>);
  // () -> void <: () -> dynamic
  Expect.isTrue(new C<void_>() is C<_>);
  // () -> int <: () -> void
  Expect.isTrue(new C<int_>() is C<void_>);
  // () -> void <: () -> int
  Expect.isFalse(new C<void_>() is C<int_>);
  // () -> void <: () -> void
  Expect.isTrue(new C<void_>() is C<void_2>);
  // () -> int <: () -> int
  Expect.isTrue(new C<int_>() is C<int_2>);
  // () -> int <: () -> Object
  Expect.isTrue(new C<int_>() is C<Object_>);
  // () -> int <: () -> double
  Expect.isFalse(new C<int_>() is C<double_>);
  // () -> int <: (int) -> void
  Expect.isFalse(new C<int_>() is C<void__int>);
  // () -> void <: (int) -> int
  Expect.isFalse(new C<void_>() is C<int__int>);
  // () -> void <: (int) -> void
  Expect.isFalse(new C<void_>() is C<void__int>);
  // (int) -> int <: (int) -> int
  Expect.isTrue(new C<int__int>() is C<int__int2>);
  // (Object) -> int <: (int) -> Object
  Expect.isTrue(new C<int__Object>() is C<Object__int>);
  // (int) -> int <: (double) -> int
  Expect.isFalse(new C<int__int>() is C<int__double>);
  // () -> int <: (int) -> int
  Expect.isFalse(new C<int_>() is C<int__int>);
  // (int) -> int <: (int,int) -> int
  Expect.isFalse(new C<int__int>() is C<int__int_int>);
  // (int,int) -> int <: (int) -> int
  Expect.isFalse(new C<int__int_int>() is C<int__int>);
  // (()->void) -> void <: ((int)->void) -> void
  Expect.isFalse(new C<inline_void_>() is C<inline_void__int>);
  // ((int)->void) -> void <: (()->void) -> void
  Expect.isFalse(new C<inline_void__int>() is C<inline_void_>);
}
